### PR TITLE
Issue#267-Change-invitation-status-using-Quartz

### PR DESCRIPTION
**Pull request by mitianin**
dev
## ZenHub

* [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009)

## Code reviewers

 - [ ] @kh0ma
 - [ ] @likeRewca
 - [ ] @Ilya-Ross
 - [ ] @AriannaLiss
 - [ ] @Illia-Shchur
 - [ ] @kizimov
 - [ ] @IhorSamoshost
 - [ ] @Vladyslav-Frolov

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/DeactivateApartmentInvitationsJob.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.quartz.JobExecutionContext;
+import org.springframework.scheduling.quartz.QuartzJobBean;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+@Slf4j
+@QuartzJobBeanAutoConfiguration(cron = "0 0 11,23 ? * *")
+public class DeactivateApartmentInvitationsJob extends QuartzJobBean {
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/DeactivateApartmentInvitationsJob.java]: As `endTime` property exists in abstract Invitation I think it makes sense to have only one job per this action. Please combine DeactivateApartmentInvitationsJob and DeactivateCooperationInvitationsJob into single job.

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java [hunk 2]
```diff
@@ -95,6 +97,7 @@ public InvitationDto saveInvitation(InvitationDto invitationDto) {
         var apartmentInvitation = mapper
                 .convert(invitationDto, ApartmentInvitation.class);
         apartmentInvitation.setStatus(InvitationStatus.PENDING);
+        apartmentInvitation.setRequestEndTime(LocalDateTime.now().plusDays(7));
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java]: Can we expose `7` as a constant?

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java [hunk 3]
```diff
@@ -139,5 +142,17 @@ public void deactivateInvitationById(Long apartmentId, Long id) {
...
     }
-}
 
+    public List<ApartmentInvitation> deactivateApartmentInvitations() {
+        List<ApartmentInvitation> deactivatedApartmentInvitations = new ArrayList<>();
+
+        for (ApartmentInvitation apartmentInvitation : apartmentInvitationRepository
+                .findAll((Specification<ApartmentInvitation>) (root, criteriaQuery, criteriaBuilder) ->
+                        getInvitationForDeactivating(root, criteriaBuilder))) {
+            apartmentInvitation.setStatus(InvitationStatus.DEACTIVATED);
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java]: What other statuses do we have? I think `overdue` will be more applicable here.

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationInvitationServiceImpl.java [hunk 4]
```diff
@@ -52,4 +53,17 @@ protected InvitationDto saveInvitation(InvitationDto invitationDto) {
...
                 .collect(Collectors.toList());
     }
+
+    public List<CooperationInvitation> deactivateCooperationInvitations() {
+        List<CooperationInvitation> deactivatedCooperationInvitations = new ArrayList<>();
+
+        for (CooperationInvitation cooperationInvitation : cooperationInvitationRepository
+                .findAll((Specification<CooperationInvitation>) (root, criteriaQuery, criteriaBuilder) ->
+                        getInvitationForDeactivating(root, criteriaBuilder))) {
+            cooperationInvitation.setStatus(InvitationStatus.DEACTIVATED);
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationInvitationServiceImpl.java]: The same:

> What other statuses do we have? I think overdue will be more applicable here.

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/DeactivateApartmentInvitationsJob.java (hunk 1 above)
Comment (mitianin) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/DeactivateApartmentInvitationsJob.java]: done

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java (hunk 2 above)
Comment (mitianin) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java]: done

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java (hunk 3 above)
Comment (mitianin) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java]: done

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationInvitationServiceImpl.java (hunk 4 above)
Comment (mitianin) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationInvitationServiceImpl.java]: done

**Review by Illia-Shchur — APPROVED** [review]
Please fix code smells.

**Review by Vladyslav-Frolov — APPROVED** [review]

**Review by kizimov — APPROVED** [review]

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/SetOverdueStatusForInvitationsJob.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import lombok.NoArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.quartz.JobExecutionContext;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.quartz.QuartzJobBean;
+import org.springframework.stereotype.Component;
+
+@Component
+@NoArgsConstructor
+@Slf4j
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/SetOverdueStatusForInvitationsJob.java]: I can't see the usages of the logger. Do we really need it?

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/SetOverdueStatusForInvitationsJob.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    @Autowired
+    private ApartmentInvitationServiceImpl apartmentInvitationService;
+
+    @Autowired
+    private CooperationInvitationServiceImpl cooperationInvitationService;
+
+    @Override
+    public void executeInternal(JobExecutionContext context) {
+        apartmentInvitationService.markAsOverdueApartmentInvitations();
+        cooperationInvitationService.markAsOverdueCooperationInvitations();
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/SetOverdueStatusForInvitationsJob.java]: Why do we have separate services for invitations?

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java [hunk 7]
```diff
@@ -187,11 +188,21 @@ public void deactivateInvitationById(Long apartmentId, Long id) {
...
-                .map(invitation -> mapper.convert(invitation, ApartmentInvitationDto.class));
+            .findAll(apartmentInvitationSpecification, PageRequest
+                .of(pageNumber - 1, pageSize))
+            .map(invitation -> mapper.convert(invitation, ApartmentInvitationDto.class));
+    }
+
+    public void markAsOverdueApartmentInvitations() {
+        var overdueApartmentInvitations = apartmentInvitationRepository
+            .findAll((Specification<ApartmentInvitation>) (root, criteriaQuery, criteriaBuilder) ->
+                getOverdueInvitation(root, criteriaBuilder));
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java]: Can we use QueryDSL for this task?

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/SetOverdueStatusForInvitationsJob.java (hunk 6 above)
Comment (kh0ma) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/SetOverdueStatusForInvitationsJob.java]: Consider injecting `Set<InvitationService>` though.

**Review by kh0ma — APPROVED** [review]
In general LGTM 👍

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/SetOverdueStatusForInvitationsJob.java (hunk 5 above)
Comment (IhorSamoshost) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/SetOverdueStatusForInvitationsJob.java]: Done

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/SetOverdueStatusForInvitationsJob.java (hunk 6 above)
Comment (IhorSamoshost) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/quartz/jobs/SetOverdueStatusForInvitationsJob.java]: Done

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java (hunk 7 above)
Comment (IhorSamoshost) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ApartmentInvitationServiceImpl.java]: Done

**Pull request merged**